### PR TITLE
This commit mostly achieves a number of objectives.

### PR DESCRIPTION
**Pull request by ProgrammerDan**
First and foremost, it introduces full 1.12 support for RecipeManager
recipe control. It achieves this through the introduction of two NMS
support classes, one to handle recipe matching for 1.12, and another to
support recipe iteration and removal for 1.12.

In Minecraft 1.12, due to advancements, Recipes have been significantly
reworked. The exact ordering of recipes is now important; specifically,
the vanilla recipe exact ordering is critical. So, the new 1.12 support
throughout RecipeManager, and especially the NMS sections, allow for
recipe "removal" simply through suppression. The recipes are modified
in-place with versions that will never match during crafting; ordering,
however, is preserved so Client-side advancements are less harmed by the
transition.

Note however, that if you do remove a recipe, clientside advancements
may have some artifacts involved when the player attempts to "pick" that
recipe from their recipe books once unlocked.

Additionally, if you replace a recipe, advancements for that recipe will
not work in the future; however, impairment will be restricted to those
recipes alone.

Future efforts might be necessary to better preserve advancements; for
instance, replace could be done "in-place" using an expansion of the
techniques used in RecipeIteratorV1_12. Removals could be similarly
hooked, suppressed by RecipeManager and otherwise untouched in the list.

I'll be exploring those options next.

Pending that, future efforts may be required to add new methods to hook
into advancements and inform the client of recipe removal. For now, this
seeks to split the difference, and allow largely backwards compatible
recipe manipulation including full support for Minecraft 1.12.

Known Issues:

* Reload does not yet work.
* Recipes from advancements normally connected to vanilla recipes removed or overridden will be lost on subsequent starts.

Thanks to Haveric for brainstorming and continued support of
RecipeManager and md_5 for the comment pointing to
necessity of preserving total ordering.

### CODE REVIEW

**Comment by ProgrammerDan** [comment]
I'll leave this for the moment, I still want to address the two known issues. I have some further ideas on that front.

**Comment by ProgrammerDan** [comment]
So for reloading, I've addressed that -- turns out I was calling the wrong Bukkit method. Corrected. As well, corrected "re-init" for 1.12 in the same area.

Additionally, I've found a workaround for "losing" advancement connection when you @override a vanilla recipe. Quite simple, actually -- give your custom recipe a name, so that the bukkit-randomnumbers name is not assigned, but your name is. Keep that name the same and it will track with future player loads.

I've updated the changelog to reflect this. It might be prudent to update elsewhere, I'm not sure.

In any case, I await your testing results.

File: src/main/java/haveric/recipeManager/RecipeManager.java [hunk 1]
```diff
@@ -175,8 +176,15 @@ public void reload(CommandSender sender, boolean check, boolean firstTime) {
...
-                Vanilla.restoreAllButSpecialRecipes();
-                Recipes.getInstance().index.putAll(Vanilla.initialRecipes);
+                if (!Version.has1_12Support()) {
+                    Vanilla.restoreAllButSpecialRecipes();
+                    Recipes.getInstance().index.putAll(Vanilla.initialRecipes);
+                } else {
+                    try {
+                        Bukkit.getServer().resetRecipes();
+                        Vanilla.init();
+                    } catch (NullPointerException npe) { /* Test Framework hates this */ }
```
Comment (haveric) [inline src/main/java/haveric/recipeManager/RecipeManager.java]: What issue are you having with the test framework here?

Comment (ProgrammerDan) [inline src/main/java/haveric/recipeManager/RecipeManager.java]: Actually, hadn't tried that specific spot without the try/catch. However, that same pattern elsewhere causes test failures as no mock was set for those methods (best guess, didn't look into it closely).

**Comment by haveric** [comment]
This is looking pretty solid. Only issue I've noticed so far though is that it kills other plugin's recipes on /rmreload

Do you know any way to get around using Bukkit.getServer().resetRecipes()? Can we undo the nullify and only remove RecipeManager's added recipes?

Alternatively, /rmreload could be disabled so that it forces a server restart on recipe changes at least until we can find a way to better handle it (or support is added to fix these issues).

**Comment by ProgrammerDan** [comment]
I've been wrestling with this too; my current thinking is the latter, that /rmreload be disabled until we can figure out a better way to handle this. I'm hesitant because I'm sure like many, /rmreload has saved my butt from silly recipe design accidents on many an occasion. 

Perhaps adding a warning, via Conversations, with a confirmation step? So that the functionality wouldn't be totally lost, but people could at least be aware of what they were about to do?

File: src/main/java/haveric/recipeManager/RecipeManager.java (hunk 1 above)
Comment (haveric) [inline src/main/java/haveric/recipeManager/RecipeManager.java]: Odd, I'm not noticing any issues with the other one after removing the try catch.

Comment (ProgrammerDan) [inline src/main/java/haveric/recipeManager/RecipeManager.java]: This block has changed homes many times in my testing branches :)

**Comment by haveric** [comment]
Is Conversations an API that I'm not aware of? Either way, that sounds like a reasonable approach. For those that don't use other plugins for recipes, no point in killing this functionality.

File: src/main/java/haveric/recipeManager/RecipeManager.java (hunk 1 above)
Comment (haveric) [inline src/main/java/haveric/recipeManager/RecipeManager.java]: I can only imagine :)  I know that feeling all too well.

If you could double check whether these are needed, that would awesome.

Comment (ProgrammerDan) [inline src/main/java/haveric/recipeManager/RecipeManager.java]: Will do!

**Comment by ProgrammerDan** [comment]
Potentially: https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/conversations/Conversation.html It "quietly" rolled out maybe 2 years ago, no fanfare, but tons of packed functionality. I've made use of it in some other plugins, it takes a bit of getting used to but is worthwhile.

**Comment by haveric** [comment]
I do vaguely recall that now. If you think it's easy enough to add, that's fine. Otherwise, I was thinking just adding messages on startup and the changelog to warn people about the changed functionality.

**Comment by ProgrammerDan** [comment]
I don't mind adding it; those other indicators are a good idea too.

**Comment by ProgrammerDan** [comment]
I got off on a tangent with myself and significantly improved the functionality of overrides -- I'll be pushing a commit after a little bit more testing where overrides to vanilla recipes don't cause a rebind of loss of advancement connectivity, they are replaced "in place" ... it's quite nice.

Also changed up how reload works ... might preserve 3rd party recipes now, or at least gives them a chance. I'll still add the warning though, as behavior is functionally "undefined".

**Comment by ProgrammerDan** [comment]
Ok. I've got the Conversation for reload set up, and I've improved /rmreload's functionality, I've also confirmed that you can "remove" and "override" recipes without messing up previously gained, or lose the ability to gain new, advancements. This should also open up the door for future improvements linking new Advancements to RecipeManager managed recipes.

So, with this, both outstanding issues remarked in my initial commit appear resolved (pending much, much more testing), the unnecessary try/catches are removed, and rmreload has a baked in warning.

Let me know what I missed! I plan to roll this into production on Saturday, hell or high water, as I'm shifting my entire pluginset to 1.12 at that time. I'll be sure to make new/future PRs as I encounter issues.

**Comment by haveric** [comment]
Testing it out right now. Two things so far: 

The Conversation check doesn't seem to cancel on other typing so the next time you type "yes", the reload still goes through

This one kills other plugin's recipes on initial load and not just after an /rmreload :(

**Comment by ProgrammerDan** [comment]
Hmm, I didn't change anything that would cause _new_ killing of plugin recipes, so that's curious. The change to reload is only when !firstTime, so doesn't run on first-load. That's a bit mystifying, perhaps the prior version did the same?

It could also be that the other plugins are mangling vanilla recipes already; that will _definitely_ create some weird situations :). What do you use, so I can test with parity?

I'll look at the Conversation, that's weird.

**Comment by ProgrammerDan** [comment]
I cannot duplicate the Conversation failure, it works as expected for me -- type "no" press enter, it gives you a cancelled message, subsequent typing of "yes" and pressing enter does not then activate the reload.

**Comment by haveric** [comment]
Ok, I'm clearly going crazy. The Conversation cancelling is now working.  Also scratch the other plugin's recipe issue.. it helps when I test the correct recipe....

**Pull request merged**